### ISSUE TITLE
App login fails when no client session present

### ISSUE THREAD

**Issue by jamiemtdwyer** [Shopify/shopify_app#524]
Login fails in the following scenario:

- Install an app A to shop X
- Open the shop X admin in incognito (you must not have app A cookies)
- Navigate to Apps => App A
- It will present you with an app login page inside an iframe
- Type in the same shop's url in the login input field
- Nothing

cc @vfonic

**Comment by vfonic** [comment]
@jamiemtdwyer thanks for opening new issue.

Actually, the login succeeds.

The bug is the following:
> Type in the same shop's url (**??? this wasn't needed before**) in the login input field

Before, when there was no session, instead of asking the user to login, the user (shop) would be automatically redirected/logged-in.

It just looks like a bad UX to ask the user, who is logged in to shop X, to type in the domain of their shop, in order to login. This should happen automatically, just like it used to do before.

**Comment by vfonic** [comment]
Maybe it's related to the new per-user authetication.

**Comment by resistorsoftware** [comment]
This is painful. I am still experiencing this exact issue and it makes it hard to swallow that Shopify App team is forcing US third-party developers to tow the line in terms of perfect session handling, when in fact the Shopify App code we are using is not quite ready for that prime-time. Would be nice if the pressure was relieved until these pesky issues are actually resolved.

**Comment by jamiemtdwyer** [comment]
Just tried to replicate this on latest master but wasn't able to. When I open Shopify in a new incognito window and navigate to the application, the app loads as expected.

![](https://screenshot.click/20-56-e0m8p-dc97v.gif)

@resistorsoftware are your replication steps the same as above? Anything I'm missing here?

**Comment by resistorsoftware** [comment]
Nope.. what I see is that if I have a tab open where both the App and Shopify have authenticated fine... if I leave for a couple of days... and come back.. Shopify tab is still A-OK.. but any attempt to use the App fails as it fails to authenticate a new session to replace the expired old one.

Yes, if I load up the App in a new incognito session all works... but it is this issue of not having a previously valid client session or whatever.. all I know is no matter the hacks people are using.. today I tried to see why the App was failing and the JS console was the same hey man, bugger off, this iframe and those sameorigin settings are not gonna fly...

**Comment by flavio-b** [comment]
@resistorsoftware, not sure if this is your exact issue, but does your app's **root action** attempts to call any resource via `ShopifyAPI::` ?

**Comment by resistorsoftware** [comment]
Nope.. in my codes to date, this is just a display only App with no API calls hooked up.

**Comment by flavio-b** [comment]
I had trouble with Shopify App **not** clearing the old session before, and my partial solution was to attempt a call, any call, on the root action to the API i.e. `ShopifyAPI::Shop.current `. This would raise an unauthorized error and the handler somewhere would clear the old session, redirecting to create a new session as expected.

**Comment by resistorsoftware** [comment]
Interesting.. has to be a way to detect stale sessions and get her going again properly.. will pay more attention to this issue this week to hopefully decide once and for all where the problem lies.

**Comment by flavio-b** [comment]
Yes, I don't know how to detect if the session is stale, other than attempting a call. By the way, the code that handles the error and clears the old session is in: `login_protection.rb`.

**Comment by joshreeves** [comment]
I was having this same problem after upgrading from `7.4.0` to `8.x.x`. After much frustration, what cleared it up for me was removing `include ShopifyApp::LoginProtection` from my `ApplicationController`. Prior to version `7.0.12` it was part of the installation instructions to add it there. And it survived all of my upgrades over the months until apparently breaking `8.0.0` and above. Hope this helps others.
